framework/sandstone.cpp

```diff
@@ -870,6 +895,26 @@ static void *thread_runner(void *arg)
                 // no rethrow
             }
         }
+
+        long double get_effective_freq_mhz(void) const
+        {
+            // For when we get bogus values (e.g. running unprivileged)
+            if (mperf_bef >= mperf_aft || aperf_bef >= aperf_aft || tsc_bef >= tsc_aft) {
+                return std::numeric_limits<long double>::quiet_NaN();
+            }
+
+            const long double secs = (time_aft.tv_sec - time_bef.tv_sec) + (time_aft.tv_nsec - time_bef.tv_nsec) / 1000000000.0L;
+            const long double tsc_freq = (tsc_aft - tsc_bef) / secs;
+            const long double perf_ratio = 1.0L * (aperf_aft - aperf_bef) / (mperf_aft - mperf_bef);
+            return tsc_freq * perf_ratio / 1000000.0L;
+        }
+
+        void get_aperf_and_mperf(uint64_t *aperf, uint64_t *mperf)
+        {
+            (void) read_msr(thread_number, 0xe8, aperf);
```

Comment:
this will not work when `taskset` or `--cpuset` option is used. OS cpu id should be passed to `(read|write)_msr`. `thread_number` is sandstone's thread id which is translated to OS cpu id as `cpu_info[thread_number].cpu_number`.

Reply:
Muchas gracias for the good catch.